File: Data Structures & Algorithms/reconstruct-flight-path/submission-2.kt
class Solution {
    fun findItinerary(tickets: List<List<String>>): List<String> {
        val adj = HashMap<String, MutableList<String>>()
        val sorted = tickets.sortedWith(Comparator<List<String>>() { c1, c2 -> 
            c1[1].compareTo(c2[1])
        })
        for (ticket in sorted) {
            if (adj[ticket[0]] == null) {
                adj[ticket[0]] = mutableListOf()
            }
            adj[ticket[0]]!!.add(ticket[1])
        }
        val res = mutableListOf("JFK")

        fun dfs(src: String): Boolean {
            if (res.size == sorted.size + 1) {
                return true
            }

            val destinations = adj[src] ?: return false
            for (i in 0 until destinations.count()) {
                val v = destinations.removeAt(i)
                res.add(v)

                if (dfs(v)) {
                    return true
                }

                destinations.add(i, v)
                res.removeLast()
            }
            return false
        }

        dfs("JFK")
        return res
    }
}

File: Data Structures & Algorithms/count-paths/submission-2.swift
class Solution {
    func uniquePaths(_ m: Int, _ n: Int) -> Int {

        var cache = [String: Int]()

        func dfs(_ r: Int, _ c: Int) -> Int {

            if r == m || c == n {
                return 0
            }

            if r == m - 1 && c == n - 1 {
                return 1
            }

            if cache["\(r)-\(c)"] != nil {
                return cache["\(r)-\(c)"]!
            }

            cache["\(r)-\(c)"] = dfs(r + 1, c) + dfs(r, c + 1)

            return cache["\(r)-\(c)"]!
        }

        return dfs(0, 0)

    }
}
